content/blog/reducing-openapi.md

```diff
@@ -0,0 +1,1550 @@
+        type: string
+        example: Simple Workflow
+        description: |
+          The event-specific title associated with the run or the run-name
+          if set, or the value of `run-name` if it is set in the workflow.
+    required:
+      - id
+      - node_id
+      - head_branch
+      - run_number
+      - display_title
+      - event
+      - status
+      - conclusion
+      - head_sha
+      - path
+      - workflow_id
+      - url
+      - html_url
+      - created_at
+      - updated_at
+      - head_commit
+      - head_repository
+      - repository
+      - jobs_url
+      - logs_url
+      - check_suite_url
+      - cancel_url
+      - rerun_url
+      - artifacts_url
+      - workflow_url
+      - pull_requests
+</code>
+</pre>
+</div>
+
+As we can see, the 'workflow-run' component is referencing other schemas in the document. This becomes important if you wish to reduce an openAPI document to a smaller size, as you would need to track all recursive references and include all components that got mentioned in the way. 
+
+Note the securitySchemas object, which specifies the authentication method used: 
+```
+securitySchemes:
+   BearerToken:
+     type: http
+     scheme: bearer
+```
+
+## How to reduce an OpenAPI Document to workable size?
+### Manual and logical approach
+Let's say we have an outline of a serverless workflow, and a very large openAPI document that exposes many system resources, much more than we intend on using.
+
+The procedure for reducing the openAPI document can be logically described as the following:
+
+1. Copy any ‘openapi’, ‘info’, ‘security’, ‘servers’ objects to the new document.    
+2. Identify all system resources that are needed for the workflow.
+3. For each resource, locate the path object  that returns the needed resource. 
+4. For each path, create a dependency tree of all components referenced by the path, and of all components referenced by those components, recursively. 
+4. Include the securitySchema of the original ‘components’ object. 
+5. Include any webhooks the application may need.  
+
+This procedure can be quite tedious when done manually, so some efforts were made to automate the process. 
+
+### Python script to reduce documents
+
+The [following script](https://github.com/parodos-dev/serverless-workflows/blob/feat/aap-db-deploy/extract_from_swagger_spec.py) can be used to extract reuced openAPI documents from larger ones. 
+The script takes an input file, and output file, and a list of tuples (path objects, http method). 
+```
+python extract.py openapi_spec.json filtered_openapi_spec.json "/apis/apps/v1/namespaces/{namespace}/deployments post" "/apis/apps/v1/namespaces/{namespace}/deployments/{name} get" "/api/v1/namespaces/{namespace}/services post" "/apis/route.openshift.io/v1/namespaces/{namespace}/routes post" "/apis/route.openshift.io/v1/namespaces/{namespace}/routes/{name} get" "/apis/route.openshift.io/v1/namespaces/{namespace}/routes/{name}/status get"
```

Comment:
this will be easier to view by breaking the lines (ATM it is rendered as a very long line).